File: src/model/musictheory/NoteAndOctave.ts
import { EqualsTrait } from '../general/EqualsTrait.js';
import { Note } from './Note.js';

export class NoteAndOctave implements EqualsTrait {
    public readonly note: Note;
    public readonly octave: number;
  
    static A_440: NoteAndOctave = new NoteAndOctave(Note.A, 4);
  
    constructor(note: Note, octave: number) {
      this.note = note;
      this.octave = octave;
    }
  
    up(semitones: number): NoteAndOctave {
      let baseNoteOctave = this.octave;
      let baseNoteNote = this.note;
  
      let newNoteSemitonesFromA = baseNoteNote.semitonesFromA + semitones;
      let newNoteOctave = baseNoteOctave;
  
      while (newNoteSemitonesFromA > 11) {
        newNoteSemitonesFromA -= 12;
        newNoteOctave += 1;
      }
  
      let newNote = Note.noteBySemitonesFromA(newNoteSemitonesFromA);
  
      return new NoteAndOctave(newNote!, newNoteOctave); // Note that the return type can be null, so use `!` as an assertion.
    }
  
    down(semitones: number): NoteAndOctave {
      let baseNoteOctave = this.octave;
      let baseNoteNote = this.note;
  
      let newNoteSemitonesFromA = baseNoteNote.semitonesFromA - semitones;
      let newNoteOctave = baseNoteOctave;
  
      while (newNoteSemitonesFromA < 0) {
        newNoteSemitonesFromA += 12;
        newNoteOctave -= 1;
      }
  
      let newNote = Note.noteBySemitonesFromA(newNoteSemitonesFromA);
  
      return new NoteAndOctave(newNote!, newNoteOctave); // Note that the return type can be null, so use `!` as an assertion.
    }
  
    toString(): string {
      return `${this.note}${this.octave}`;
    }
  
    getSemitonesFromA0(): number {
      return this.note.semitonesFromA + this.octave * 12;
    }
  
    isHigherThan(other: NoteAndOctave): boolean {
      return this.getSemitonesFromA0() > other.getSemitonesFromA0();
    }
  
    isLowerThan(other: NoteAndOctave): boolean {
      return this.getSemitonesFromA0() < other.getSemitonesFromA0();
    }

    equals(other: any): boolean {
        if (!(other instanceof NoteAndOctave)) {
          return false;
        }
        
        return this.note.equals(other.note) 
          && this.octave === other.octave;
      }

  }

File: ChordVoicingsCreatorJava/src/main/java/com/campoy/chord/voicing/creator/model/musictheory/NoteAndOctave.java
package com.campoy.chord.voicing.creator.model.musictheory;

import lombok.Data;

@Data
public class NoteAndOctave {

    private final Note note;
    
    private final int octave;
    
    public static NoteAndOctave A_440 = new NoteAndOctave(Note.A, 4);
    
//    @Override
//    public boolean equals(Object obj) {
//        
//        if( !( obj instanceof NoteAndOctave) ) {
//            return false;
//        }
//        
//        NoteAndOctave objOctavatedNote = (NoteAndOctave)obj;
//        boolean equalNote = this.getNote() == objOctavatedNote.getNote();
//        boolean equalOctave = this.getOctave() == objOctavatedNote.getOctave();
//        
//        if(equalNote && equalOctave) {
//            return true;
//        }
//        else {
//            return false;
//        }
//    }
    

    public NoteAndOctave up(int semitones) {
        
        int baseNoteOctave = this.getOctave();
        Note baseNoteNote = this.getNote();
        
        int newNoteSemitonesFromA  = baseNoteNote.getSemitonesFromA() + semitones;
        int newNoteOctave = baseNoteOctave;
        
        while(newNoteSemitonesFromA > 11 ) {
            newNoteSemitonesFromA -= 12;
            newNoteOctave += 1;
        }
        
        Note newNote = Note.noteBySemitonesFromA(newNoteSemitonesFromA);
        
        return new NoteAndOctave(newNote, newNoteOctave);
        
    }
    
    public NoteAndOctave down(int semitones) {
        
        int baseNoteOctave = this.getOctave();
        Note baseNoteNote = this.getNote();
        
        int newNoteSemitonesFromA  = baseNoteNote.getSemitonesFromA() - semitones;
        int newNoteOctave = baseNoteOctave;
        
        while(newNoteSemitonesFromA < 0 ) {
            newNoteSemitonesFromA += 12;
            newNoteOctave -= 1;
        }
        
        Note newNote = Note.noteBySemitonesFromA(newNoteSemitonesFromA);
        
        return new NoteAndOctave(newNote, newNoteOctave);
        
    }
    
    @Override
    public String toString() {
        return "" + note + octave;
    }
    
    public int getSemitonesFromA0() {
        return note.getSemitonesFromA() + octave * 12;
    }    
    
    public boolean isHigherThan(NoteAndOctave other) {
        return getSemitonesFromA0() > other.getSemitonesFromA0();        
    }
    
    public boolean isLowerThan(NoteAndOctave other) {
        return getSemitonesFromA0() < other.getSemitonesFromA0();        
    }
    
}
